### DISCUSSION TITLE
Export unicode text in dxf file into pdf

### DISCUSSION THREAD

**Discussion by ebrahimraeyat** [mozman/ezdxf#967] [Q&A]
Hello. First thanks a lot for such interesting project. I use it in my workflow.

I have some persian text in my dxf file and wanted to export it to pdf with backend.

but when I export it, it did not show unicode text correctly.

Also I identify fonts folder:

``` python

import ezdxf
ezdxf.options.set(ezdxf.options.CORE, "support_dirs" , "G:\Program Files\Autodesk\AutoCAD 2020\Fonts")
ezdxf.options.write_file()
```

![image](https://github.com/mozman/ezdxf/assets/8196112/7c8a9b3f-98f9-4852-8770-cc28bfda295c)

![image](https://github.com/mozman/ezdxf/assets/8196112/4c6d7859-1c59-4188-bbe2-e65cd17b1426)

**Comment by mozman** [comment]
Hi!

You did everything right, but you didn't rebuild the font cache, see [Rebuild Font Cache](https://ezdxf.mozman.at/docs/tools/fonts.html#rebuilding-the-font-cache).

The font cache is created once when `ezdxf` is first started and is not automatically updated afterwards.

You should also create a permanent configuration file in your home directory, otherwise you will rebuild the font cache on every launch, which defeats the purpose of the cache:

```shell
ezdxf config --home
```
The config file is located at: `C:\users\<username>\.cache\ezdxf\ezdxf.ini`

You can also create the config file in your current work directory:

```shell
ezdxf config --print > ezdxf.ini
```

see also [Config Files](https://ezdxf.mozman.at/docs/options.html#config-files)

And also note, the text rendering engine was not designed with active support for Right-To-Left languages, as I have no idea how that works.

**Comment by ebrahimraeyat** [comment]
Thanks @mozman  for your quick and complete response. 
I executed this command:
```python
import ezdxf
from ezdxf.fonts import fonts
fonts.build_system_font_cache()
```
 and gives me this errors:

```

'name' table stringOffset incorrect. Expected: 222; Actual: 224
'name' table stringOffset incorrect. Expected: 294; Actual: 210
skipping malformed name record #17
skipping malformed name record #18
skipping malformed name record #19
skipping malformed name record #20
skipping malformed name record #21
skipping malformed name record #22
skipping malformed name record #23
....
'name' table stringOffset incorrect. Expected: 294; Actual: 210
skipping malformed name record #17
skipping malformed name record #18
skipping malformed name record #19
skipping malformed name record #20
skipping malformed name record #21
skipping malformed name record #22
skipping malformed name record #23
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "G:\freecad\git\build\bin\lib\site-packages\ezdxf\fonts\fonts.py", line 126, in build_system_font_cache
    build_font_manager_cache(_get_font_manager_path())
  File "G:\freecad\git\build\bin\lib\site-packages\ezdxf\fonts\fonts.py", line 326, in build_font_manager_cache
    font_manager.build()
  File "G:\freecad\git\build\bin\lib\site-packages\ezdxf\fonts\font_manager.py", line 409, in build
    self.scan_all(dirs + list(options.support_dirs))
  File "G:\freecad\git\build\bin\lib\site-packages\ezdxf\fonts\font_manager.py", line 417, in scan_all
    self.scan_folder(Path(folder).expanduser())
  File "G:\freecad\git\build\bin\lib\site-packages\ezdxf\fonts\font_manager.py", line 431, in scan_folder
    font_face = get_ttf_font_face(file)
  File "G:\freecad\git\build\bin\lib\site-packages\ezdxf\fonts\font_manager.py", line 463, in get_ttf_font_face
    style = record.string.decode(record.getEncoding())
  File "G:\freecad\git\build\bin\lib\encodings\utf_16_be.py", line 16, in decode
    return codecs.utf_16_be_decode(input, errors, True)
UnicodeDecodeError: 'utf-16-be' codec can't decode byte 0x72 in position 6: truncated data
```

It created a json file in ```C:\Users\{username}\.cache\ezdxf\font_manager_cache.json``` with this content:

```json
{
  "version": 1,
  "font-faces": [
    [
      "C:\\Users\\ebrahim\\AppData\\Local\\Microsoft\\Windows\\Fonts\\2ASEMAN.TTF",
      "2  Aseman",
      "Regular",
      400,
      5
    ],
    [
      "C:\\Users\\ebrahim\\AppData\\Local\\Microsoft\\Windows\\Fonts\\2ASEMANI.TTF",
      "2  Aseman",
      "Italic",
      400,
      5
    ],
....
```

Then I created pdf again and it only shows the rectangle around the text:

![image](https://github.com/mozman/ezdxf/assets/8196112/d62817e5-f3b8-48b3-b8a8-d35fb590402d)

![image](https://github.com/mozman/ezdxf/assets/8196112/7e69be03-e4c4-4aa2-a5c4-95e6fd77661b)

my configuration:

```python
cfg = config.Configuration(
        background_policy=config.BackgroundPolicy.CUSTOM,
        custom_bg_color="#ffffff00",
        min_lineweight=.1,
        lineweight_scaling=0,
        color_policy=config.ColorPolicy.BLACK,
    )
```

Thanks again.

**Reply by mozman** [reply]
"Your" fonts are encoded in way that they throw an exception. There is nothing you can do until you install the next version of `ezdxf` that contains a fix for this specific bug. However, this does not mean that the next version will be able to render your file.

I'm wondering why this error didn't occur before, since the error is caused by a TTF font. And the first build of the font cache didn't throw an exception. After starting `ezdxf` for the first time, new TTF fonts must have been installed.

**Reply by mozman** [reply]
I released the next beta version v1.1.3b1 so you can test the fix.

**Comment by ebrahimraeyat** [comment]
Wow! Thanks.
I examined it with new version, But nothing happened. I run the rebuild font cache command again and this is output:

```
name' table stringOffset incorrect. Expected: 222; Actual: 224
'name' table stringOffset incorrect. Expected: 294; Actual: 210
skipping malformed name record #17
skipping malformed name record #18
skipping malformed name record #19
skipping malformed name record #20
skipping malformed name record #21
skipping malformed name record #22
skipping malformed name record #23
'name' table stringOffset incorrect. Expected: 294; Actual: 210
...
skipping malformed name record #23
cannot open font 'G:\Program Files\Autodesk\AutoCAD 2020\Fonts\OldGlory.otf': 'utf-16-be' codec can't decode byte 0x72 in position 6: truncated data
'name' table stringOffset incorrect. Expected: 294; Actual: 210
skipping malformed name record #17
...
skipping malformed name record #23
cannot open font 'G:\Program Files\Autodesk\AutoCAD 2020\Fonts\ZWAdobeF.TTF': Not a TrueType or OpenType font (bad sfntVersion)
```
Then when I export **dxf** file to pdf this error appears:

```
DirectWrite: CreateFontFaceFromHDC() failed (Indicates an error in an input file such as a font file.) for QFontDef(Family="8514oem", pointsize=10, pixelsize=20, styleHint=5, weight=75, stretch=100, hintingPreference=0) LOGFONT("8514oem", lfWidth=0, lfHeight=-20) dpi=120
20:01:32  DirectWrite: CreateFontFaceFromHDC() failed (Indicates an error in an input file such as a font file.) for QFontDef(Family="8514oem", pointsize=10, pixelsize=20, styleHint=5, weight=50, stretch=100, hintingPreference=0) LOGFONT("8514oem", lfWidth=0, lfHeight=-20) dpi=120
```

BTW, this time the English text don't appear in most cases:

before:

![image](https://github.com/mozman/ezdxf/assets/8196112/c276a195-9d45-4985-bf5b-e3f2dbba8bdc)

with new version:

![image](https://github.com/mozman/ezdxf/assets/8196112/6e0a661c-cb52-4ae1-8035-3d342619fbef)

Thanks a lot.

**Reply by mozman** [reply]
I can't help you because I can't replicate your system and your installed fonts. You have many fonts with issues that I don't understand and cannot fix. Sorry.

**Reply by mozman** [reply]
You can upload the drawing as a zip-archive and I will look into it to find  the problematic fonts and see if the drawing will be rendered with replacement fonts on my system.

**Comment by ebrahimraeyat** [comment]
Yes, you are right. I attached my dxf file with naskhd.shx persian font:

[persian_unicode_text.zip](https://github.com/mozman/ezdxf/files/13404548/persian_unicode_text.zip)

Thanks again.

**Comment by mozman** [comment]
The missing text has the text style `SAZE_STYLE` and uses the `romansp.shx` font which I don't have (I have collected a lot) and yours seems to be incompatible to `ezdxf`  or broken - maybe you can send me your `romansp.shx` font to <confidential@mozman.at>

You can replace the `romansp.shx` by `romans.shx`:
```Python
import ezdxf

doc = ezdxf.readfile("persian_text.dxf")
style = doc.styles.get("SAZE_STYLE")
style.dxf.font = "romans.shx"
doc.saveas("fixed_persian_text.dxf ")
```

There is another error without an easy fix. When I add the `NASKHD.SHX` font to my fonts, `ezdxf` can't use the font and doesn't even find a replacement font - this is an error in `ezdxf`.

![image](https://github.com/mozman/ezdxf/assets/4230966/a00568f6-b990-4303-beb1-04e60252f1b9)

**Reply by ebrahimraeyat** [reply]
My Persian font style is ```civillir``` with font name ```naskhd.shx```. The concrete column font is in Latin.
BTW i will send the fonts. Thanks.

**Reply by mozman** [reply]
It was a terrible beginner error 😁, but as expected the rendering is bad:

![image](https://github.com/mozman/ezdxf/assets/4230966/8747c3b8-62c5-4377-a21f-1958ca4dc46f)
